### PR TITLE
feat: Restrict the total number of 'Running' workspaces on a cluster

### PR DESCRIPTION
**Pull request by tolusha**
### What does this PR do?
feat: Restrict the total number of 'Running' workspaces on a cluster

### Screenshot/screencast of this PR

screenshot-192_168_59_253_nip_io-2024_08_29-15_47_45

### What issues does this PR fix or reference?
https://github.com/eclipse-che/che/issues/23081

### Is it tested? How?

1. Deploy Che on Minikube
2. `kubectl patch -n eclipse-che "checluster/eclipse-che" --type=json -p="[{"op": "replace", "path": "/spec/components/dashboard/deployment", "value": {containers: [{image: "quay.io/eclipse/che-dashboard:pr-1173", name: che-dashboard}]}}]"`
3. `kubectl patch -n eclipse-che "checluster/eclipse-che" --type=json -p="[{"op": "replace", "path": "/spec/devEnvironments/maxNumberOfRunningWorkspacesPerCluster", "value": 1}]"`
4. Login as user1, start a workspace
5. Login as user2, start a workspace and got error message

#### Release Notes

```Introduced a new configuration option, `maxNumberOfRunningWorkspacesPerCluster`, to limit the total number of 'Running' workspaces on a cluster. When the limit is reached, users attempting to start additional workspaces will receive an error message.

#### Docs PR

TODO

### CODE REVIEW

File: packages/dashboard-frontend/src/components/WorkspaceProgress/CommonSteps/CheckRunningWorkspacesLimit/__tests__/index.spec.tsx [hunk 1]
```diff
@@ -355,6 +355,56 @@ describe('Common steps, check running workspaces limit', () => {
...
+      storeBuilder = new FakeStoreBuilder()
+        .withDevWorkspaces({
+          workspaces: [targetDevworkspace, runningDevworkspace1],
+        })
+        .withClusterConfig({
+          runningWorkspacesLimit: 10,
+        });
+    });
+
+    test.only('alert notification', async () => {
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/components/WorkspaceProgress/CommonSteps/CheckRunningWorkspacesLimit/__tests__/index.spec.tsx:373]:
```suggestion
    test('alert notification', async () => {
```

File: packages/dashboard-frontend/src/components/WorkspaceProgress/CommonSteps/CheckRunningWorkspacesLimit/__tests__/index.spec.tsx [hunk 2]
```diff
@@ -355,6 +355,56 @@ describe('Common steps, check running workspaces limit', () => {
...
+            callback: expect.any(Function),
+          }),
+        ],
+      });
+      await waitFor(() => expect(mockOnError).toHaveBeenCalledWith(expectAlertItem));
+      expect(mockOnNextStep).not.toHaveBeenCalled();
+      expect(mockOnRestart).not.toHaveBeenCalled();
+    });
+
+    test.only('start a workspace', async () => {
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/components/WorkspaceProgress/CommonSteps/CheckRunningWorkspacesLimit/__tests__/index.spec.tsx:395]:
```suggestion
    test('start a workspace', async () => {
```

File: packages/dashboard-backend/src/devworkspaceClient/services/__tests__/devWorkspaceClusterApi.spec.ts [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (olexii4) [inline packages/dashboard-backend/src/devworkspaceClient/services/__tests__/devWorkspaceClusterApi.spec.ts:13]:
```suggestion
```

Comment (olexii4) [reply packages/dashboard-backend/src/devworkspaceClient/services/__tests__/devWorkspaceClusterApi.spec.ts:13]: @tolusha Do we really need it?

**Review by akurinnoy — APPROVED** [review]

**Review by olexii4 — APPROVED** [review]
LGTM

**Pull request merged**